tar/test/test_option_n.c

```diff
@@ -58,4 +60,91 @@ DEFINE_TEST(test_option_n)
 	assertIsDir("d1", umasked(0755));
 	assertFileNotExists("d1/file1");
 	assertChdir("..");
+
+	/*
+	 * Create a test archive with the following content:
+	 * d1/
+	 * d1/file1
+	 * d1/file2
+	 * file3
+	 * d2/file4
+	 *
+	 * Extracting uses the same code as listing and thus does not
+	 * get tested separately. This also covers the
+	 * archive_match_include_directories_recursively()
+	 * API.
+	 */
+	assertMakeFile("d1/file2", 0644, "d1/file2");
+	assertMakeFile("file3", 0644, "file3");
+	assertMakeDir("d2", 0755);
+	assertMakeFile("d2/file4", 0644, "d2/file4");
+	assertEqualInt(0,
+	    systemf("%s -cnf partial-archive.tar d1 d1/file1 d1/file2 file3 d2/file4 >c.out 2>c.err", testprog));
```

Comment:
I would like to extend this test slightly to verify that this does the expected thing when you have competing inclusions/exclusions.  You'll need a deeper directory tree, something like this:

* d1
* d1/d2a
* d1/d2a/d3
* d1/d2b

Test cases:
1. Include "d1", exclude "d1/d2a", recursive matching enabled
2. Exclude "d1", include "d1/d2a", recursive matching enabled
3. Include "d1", exclude "d1/d2a", recursive matching disabled
4. Exclude "d1", include "d1/d2a", recursive matching disabled